CRM/Birthdays/Mailer.php

```diff
@@ -0,0 +1,121 @@
+ */
+
+use Civi\API\Exception\UnauthorizedException;
+use Civi\Api4\Activity;
+use CRM_Birthdays_ExtensionUtil as E;;
+
+class CRM_Birthdays_Mailer
+{
+    private int $template_id;
+    private string $email_address_from;
+
+    /**
+     * @throws Exception
+     */
+    public function __construct()
+    {
+        $this->template_id = Civi::settings()->get(CRM_Birthdays_Form_Settings::BIRTHDAYS_MESSAGE_TEMPLATE);
+        if (empty($this->template_id)) {
+            throw new Exception('Birthdays: Message template not found. Please set a template in birthday settings');
+        }
+
+        $email_id = Civi::settings()->get(CRM_Birthdays_Form_Settings::BIRTHDAYS_SENDER_EMAIL_ADDRESS_ID);
+        $this->email_address_from = $this->get_sender_email_address_from_id($email_id);
+        if (empty($this->email_address_from)) {
+            throw new Exception('Birthdays: Pre selected outgoing email not found. Please set am outgoing email address in birthday settings');
+        }
+    }
+
+    /**
+     * @param $contacts
+     * @param $write_activity
+     * @return int error count
+     * @throws API_Exception
+     * @throws CRM_Core_Exception
+     * @throws UnauthorizedException
+     */
+    public function send_mails_and_write_activity($contacts, $write_activity): int
+    {
+        $error_count = 0;
+        foreach ($contacts as $contact_id => $contact_info) {
+            try {
+                $this->send_mail($contact_id, $this->email_address_from, $contact_info['email'], $this->template_id);
+                if ($write_activity) $this->create_activity($contact_id, E::ts('Successful birthday greeting mail'), E::ts('Successful birthday greeting mail! Template ID %1 has been used.', [$this->template_id]));
+            } catch (Exception $exception) {
+                if ($write_activity) $this->create_activity($contact_id, E::ts('FAILED birthday greeting mail'), E::ts("Failed sending an birthday greeting mail with template ID nr %1. Error: %2", [$this->template_id, $exception]));
+                ++$error_count;
+            }
+        }
+        return $error_count;
+    }
+
+    /**
+     * @throws CRM_Core_Exception
+     * @throws Exception
+     */
+    private function send_mail($contact_id, $from_email_address, $to_email_address, $template_id): void
+    {
+        try {
+            civicrm_api3('MessageTemplate', 'send', [
+                'check_permissions' => 0,
+                'id'                => $template_id,
+                'to_name'           => civicrm_api3('Contact', 'getvalue', ['id' => $contact_id, 'return' => 'display_name']),
+                'from'              => trim($from_email_address),
+                'contact_id'        => $contact_id,
+                'to_email'          => trim($to_email_address),
+            ]);
+        } catch (Exception $exception) {
+            throw new Exception("Birthdays: MessageTemplate exception: $exception");
```

Comment:
Using `E::LONG_NAME` as an exception message prefix might be better than just `Birthdays`. This should then apply to all exception messages.